### PR TITLE
(4/n) Data Refactor - Finetuning Scripts

### LINKED ISSUE #951: Trainer args consistency
For consistency, should it be `io.destination_path` (or perhaps even better `io.out_dir`) in 

```bash
python scripts/prepare_alpaca.py \
   --destination_path data/alpaca \
   --checkpoint_dir checkpoints/google/gemma-2b


python finetune/lora.py \
   --io.train_data_dir data/alpaca \
   --checkpoint_dir checkpoints/google/gemma-2b
```

### PR DESCRIPTION
**Pull request by awaelchli**
Fixes https://github.com/Lightning-AI/lit-gpt/issues/954
Fixes #951

### CODE REVIEW

File: finetune/full.py [hunk 1]
```diff
@@ -155,14 +153,15 @@ def fit(
         f" {model.max_seq_length} and context length is {model.config.block_size}"
```
Comment (rasbt) [inline finetune/full.py:153]: Should 

> get_longest_seq_length(train_data)

be

> get_longest_seq_length(train_dataloader)

?

**Comment by rasbt** [comment]
When running with 1 epoch and an epoch size of 50000 on Alpaca (default 64 global batch size and microbatch size of 1), that's currently how the learning rate looks like during the course of the training:

plot (1)


This looks good and is exactly what I would expect.

When I reduced the training epoch size to 1000 and adjusted the warmup steps 100 -> 10, then it seems to be doing something weird:

plot (4)

I think we need to adjust the code so that the scheduler steps here

```python
    steps_per_epoch = len(train_dataloader) // train.gradient_accumulation_iters(devices)
    lr_max_steps = train.epochs * steps_per_epoch
```

are perhaps computed by the `epoch_size * train.epochs` so that this doesn't happen? We can use  `len(train_dataloader) // train.gradient_accumulation_iters(devices)` then if `train.epoch_size` is `None`. What do you think?

File: finetune/full.py [hunk 2]
```diff
@@ -270,51 +265,24 @@ def validate(
...
+    datamodule.setup()
+    train_dataloader = datamodule.train_dataloader()
+    val_dataloader = datamodule.val_dataloader()
+    train_dataloader, val_dataloader = fabric.setup_dataloaders(train_dataloader, val_dataloader)
+    return train_dataloader, val_dataloader
+
+
 def get_longest_seq_length(data: List[Dict]) -> Tuple[int, int]:
     # find out the minimum max_seq_length required during fine-tuning (saves memory!)
     lengths = [len(d["input_ids"]) for d in data]
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply finetune/full.py:290]: I agree, we definitely need to remove this requirement now. 

Regarding the fast-dev-run experience, would it be ok to control this via max_steps / max_iters? 
It would be slightly easier to implement it this way and perhaps more intuitive?

File: finetune/full.py [hunk 3]
```diff
@@ -21,24 +23,25 @@
...
+    CycleIterator,
 )
-from scripts.prepare_alpaca import generate_prompt
 
 
 def setup(
     precision: Optional[str] = None,
     devices: int = 1,
     resume: Union[bool, Path] = False,
+    data: Optional[LitDataModule] = None,
```
Comment (awaelchli) [inline finetune/full.py:42]: Found a small bug in jsonargparse: https://github.com/omni-us/jsonargparse/issues/460
So for now, using None as sentinel (maybe that's better anyway)

File: finetune/adapter.py [hunk 4]
```diff
@@ -149,34 +150,36 @@ def fit(
...
 
+    train_iterator = CycleIterator(train_dataloader)
     throughput = ThroughputMonitor(fabric, window_size=50)
     step_count = 0
+    iter_num = 0
     total_lengths = 0
     total_t0 = time.perf_counter()
 
-    for iter_num in range(1, train.max_iters(devices) + 1):
+    while step_count < (train.max_steps or float("inf")) and train_iterator.epoch < train.epochs:
```
Comment (carmocca) [inline finetune/adapter.py:177]: Can you `or train.max_steps` outside the loop?

File: finetune/adapter.py [hunk 5]
```diff
@@ -316,8 +294,8 @@ def validate_args(io: IOArgs, train: TrainArgs, eval: EvalArgs) -> None:
             if getattr(args, name) is not None:
                 issues.append(f"{__file__} doesn't support the {name!r} argument. This is set in {args}")
     required = [
-        (io, ["checkpoint_dir", "train_data_dir", "val_data_dir"]),
-        (train, ["epoch_size", "epochs"]),
+        (io, ["checkpoint_dir"]),
+        (train, ["epochs"]),
```
Comment (carmocca) [inline finetune/adapter.py:298]: Do we need to update this logic to check that one of "epochs" or "max_steps" is set?

File: finetune/adapter.py [hunk 6]
```diff
@@ -131,10 +134,8 @@ def main(fabric: L.Fabric, devices: int, seed: int, config: Config, io: IOArgs,
     # strict=False because missing keys due to Adapter weights not contained in state dict
     load_checkpoint(fabric, model, checkpoint_path, strict=False)
 
-    fabric.seed_everything(1337 + fabric.global_rank)
```
Comment (carmocca) [inline finetune/adapter.py:134]: You removed this because the dataloaders seed themselves, right? But isn't the new version missing the `global_rank` increase?

File: pretrain/openwebtext.py
Comment (carmocca) [inline pretrain/openwebtext.py:1]: This is what you said will be broken for now?

File: pretrain/redpajama.py
Comment (carmocca) [inline pretrain/redpajama.py:1]: This is removed because there's no more PackedDataset and redpajama is supported through the datamodule?

File: tests/conftest.py [hunk 7]
```diff
@@ -74,6 +75,30 @@ def mock_tockenizer():
     return MockTokenizer()
 
 
+@pytest.fixture()
```
Comment (carmocca) [inline tests/conftest.py:78]: You moved this because you need these fixtures outside of `tests/data`?

**Review by carmocca — COMMENTED** [review]
My comments would apply to all the other files too

File: finetune/adapter.py (hunk 6 above)
Comment (awaelchli) [reply finetune/adapter.py:134]: The seeding is for the dataset/dataloader yes and is now handled in the module.
The global rank thing is not necessary anymore, because `fabric.setup_dataloaders()` will configure the distributed sampler in the DataLoader.

File: pretrain/openwebtext.py
Comment (awaelchli) [reply pretrain/openwebtext.py:1]: Not broken in the sense that there is a Python error. Just that it doesn't use the new data abstraction and who even knows if this even trains properly. It does not make sense to work on this script if it gets removed and replaced with tinyllama.py anyway.

File: pretrain/redpajama.py
Comment (awaelchli) [reply pretrain/redpajama.py:1]: There is no need for packed dataset anymore. Training with Redpajama is obsolete now that we have slimpajama.

File: finetune/adapter.py (hunk 4 above)
Comment (awaelchli) [reply finetune/adapter.py:177]: Actually, I don't know what you mean. Can you clarify

Comment (carmocca) [reply finetune/adapter.py:177]: I didn't write what I was thinking, edited

Comment (awaelchli) [reply finetune/adapter.py:177]: Like `train.max_steps = train.max_steps or float("inf")` for example?

Comment (carmocca) [reply finetune/adapter.py:177]: Yes but do not modify the config, we should treat it as immutable. Save it into a variable

**Review by carmocca — APPROVED** [review]
